### PR TITLE
Fixed: Duplicating a file with an appended filename can cause an Editor deadlock

### PR DESCRIPTION
**Pull request by jjjoness**
Signed-off-by: John Jones-Steele <82226755+jjjoness@users.noreply.github.com>

## What does this PR do?

Fixes critical bug #13486

## How was this PR tested?
Tried various previously failing filenames.

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp [hunk 1]
```diff
@@ -2334,7 +2334,7 @@ namespace AZ::StringFunc
                 AZStd::string originalFname;
                 do
                 {
-                    originalFname = AZStd::string(stem.Native()) + suffix + AZStd::to_string(value++);
+                    originalFname = AZStd::string(stem.Native()) + suffix + AZStd::to_string(value++) + extension.Native().data();
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: AZStd::format might still be nicer here but otherwise looks good

**Review by hultonha — APPROVED** [review]

**Review by cgalvan — APPROVED** [review]

**Review by dtamkin1 — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: This function ([MakeUniqueFilenameWithSufix](https://github.com/o3de/o3de/pull/10938)) was originally added without someone from Sig-Core signing off on it.
I have a problem with how this function is authored.

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: I would author the function as follows.

```c++
AZ::IO::FixedMaxPath MakeUniqueFilename(const AZ::IO::PathView& basePath)
{
    AZ::IO::FixedMaxPath outFilePath(basePath);
    if (!AZ::IO::FileIOBase::GetInstance()->Exists(outFilePath.c_str()))
    {
        return outFilePath;
    }

    AZ::IO::PathView extension = basePath.Extension();
    AZ::IO::PathView stem = basePath.Stem();
    int value = 1;
    do
    {
        auto uniqueFilename =  AZ::IO::FixedMaxPathString::format("%.*s%d.%.*s", AZ_STRING_ARG(stem.Native()), value, AZ_STRING_ARG(extension.Native()));
        outFilePath.ReplaceFilename(AZ::IO::PathView(uniqueFilename));
    } while (AZ::IO::FileIOBase::GetInstance()->Exists(outFilePath.c_str()));

    return outFilePath;
}
```

Comment (jjjoness) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: Thanks for your input, I'll make your changes now and test them.

Comment (jjjoness) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: @lumberyard-employee-dm I need to spend a bit more time on this. This function doesn't have all the functionality of MakeUniqueFilenameWithSuffix.

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: Well this function is used in three locations(1 in AssetBrowserTreeView.cpp and 2 in SystemComponent.cpp).
The function doesn't need to pass in separate portions of the path, but instead a single path.
The use where this is attempting to fix the issue is in [AssetBrowserTreeView.cpp](https://github.com/o3de/o3de/blob/development/Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Views/AssetBrowserTreeView.cpp#L747-L751)

That code should be
```
                AZ::IO::Path oldPath = entry->GetFullPath();
                AZ::IO::FixedMaxPath newPath = AZ::StringFunc::MakeUniqueFilename(oldPath.Native() + "-copy");
                QFile::copy(oldPath.c_str(), newPath.c_str());
```

Comment (jjjoness) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp:2337]: No, that doesn't work. You would end up with a file like "test.png-copy". I'll fix it to use most of your code and post it when tested.

Comment (jjjoness) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp]: @lumberyard-employee-dm Made changed in line with your comments. Could you take a look please?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/StringFunc/StringFunc.cpp]: Will do

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**